streams/src/main/java/org/apache/kafka/streams/state/internals/ValueAndTimestampDeserializer.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.LongDeserializer;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+
+import java.util.Arrays;
+import java.util.Map;
+
+class ValueAndTimestampDeserializer<V> implements Deserializer<ValueAndTimestamp<V>> {
+    public final Deserializer<V> valueDeserializer;
+    private final Deserializer<Long> timestampDeserializer;
+
+    ValueAndTimestampDeserializer(final Deserializer<V> valueDeserializer) {
+        this.valueDeserializer = valueDeserializer;
+        timestampDeserializer = new LongDeserializer();
+    }
+
+    @Override
+    public void configure(final Map<String, ?> configs,
+                          final boolean isKey) {
+        valueDeserializer.configure(configs, isKey);
+        timestampDeserializer.configure(configs, isKey);
+    }
+
+    @Override
+    public ValueAndTimestamp<V> deserialize(final String topic,
+                                            final byte[] data) {
+        if (data == null) {
+            return null;
+        }
+        final long timestamp = timestampDeserializer.deserialize(topic, Arrays.copyOfRange(data, 0, 8));
```

Comment:
Using `Arrays.copyRange` here -- on other places, I uses `System.arraycopy()` -- other code uses `ByteBuffer` -- not sure which one is the best. Any insights?

Reply:
Interesting! I was unaware of this method.

The range check is just making sure that the "end" index isn't after the "start" index. This is probably just to give a nice error message, since it would otherwise result in a negative "length" argument and throw a different runtime exception (`java.lang.ArrayIndexOutOfBoundsException`)

It seems like the real advantage is the compile-time type safety it gives, since there's nothing preventing you from using `System.arrayCopy` to copy from an `int[]` into a `String[]`. You'd get a runtime exception (`java.lang.ArrayStoreException`). Using `Arrays.copyOfRange`, you get a compile-time check that your types are right.

Plus, it creates the destination array for you, which is also handy.

Since Java seems to lack a structural array slice (I.e., get a new array that is actually a view on a backing array), you have to make a copy to slice the array, so the current method seems appropriate.

`ByteBuffer` actually does let you do a structural slice, but afaict, any mechanism that gets an array from a ByteBuffer either gives you the whole (unsliced) backing array or makes a copy any way. If the deserializer accepted a ByteBuffer, we could do these deserializations without making array copies, but as it is, it seems like the Deserializer interface is painting us into a corner by requiring arrays as input.

The reason (I think) that we use ByteBuffer sometimes is that its builder methods make it easy to pack and unpack data into and out of arrays. This can be an error prone process in general, so it's sometimes nice. But what you're doing is pretty straightforward, and as long as the serializer and deserializer are symmetrical, it's going to be fine.

Reply:
What is your conclusion on what we should use?

Reply:
I'd vote to go with `Arrays.CopyOfRange` since its extra check logic does not seem to incur a huge overhead, so although this is on the critical code path it may still not a big perf regression.

Reply:
As discussed in person, we agreed on using `ByteBuffer`.